### PR TITLE
feat: scaling down pods before starting the Longhorn to OpenEBS/Rook migration.

### PR DESCRIPTION
**Pull request by ricardomaraschini**
#### What this PR does / why we need it:

This PR introduces a new command flag to the Kurl binary, allowing us to gracefully scale down pods using Longhorn volumes **prior** to migration (before running `pvmigrate`). By incorporating this flag into the Rook and OpenEBS pre-init process, we aim to scale down Pods using a more stable cluster (before start to bump things around and generate Longhorn instability). My testing has shown a noticeable improvement, as there have been no instances of Pods stuck in the "Terminating" state anymore.

However, in the event that some pods do become stuck during pre-flight, the migration will fail and may leave the cluster with some Pods down and Longhorn volumes scaled down. This is by design as it allows us to take action and manually restart Kubelet in the affected nodes. In case of failure a helpful message is also displayed with instructions on how to bring the scaled-down Pods and Volumes back up using another flag (also introduced by this PR) in the Kurl binary.

Migrations from Longhorn to Rook have been **SUSPENDED** due to failures encountered during testing on a single node cluster with 8 CPUs and 32GB of RAM. The failures were characterized by issues with etcd and the controller manager, which have been linked to a specific Kernel parameter (as previously discussed on Slack). 

#### Which issue(s) this PR fixes:
Fixes #65977

### CODE REVIEW

File: pkg/cli/longhorn.go [hunk 1]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
+	for _, obj := range objects {
+		if err := scaleDownObject(ctx, cli, obj); err != nil {
+			return err
+		}
+	}
+	log.Print("Pods using Longhorn volumes have been scaled down.")
+	return nil
+}
+
+// scalePrometheus scales dprometheus machinery.
```
Comment (emosbaugh) [inline pkg/cli/longhorn.go:249]:
```suggestion
// scalePrometheus scales prometheus statefulsets to the number of provided replicas.
```

File: pkg/cli/longhorn.go [hunk 2]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
 	}
 }
 
+// scaleUpPodsUsingLonghorn scales up any deployment or statefulset that has been previously
+// scaled down by scaleDownPodsUsingLonghorn. uses the default annotation used by pvmigrate.
+func scaleUpPodsUsingLonghorn(ctx context.Context, cli client.Client) error {
+	if err := scaleEkco(ctx, cli, 1); err != nil {
+		return fmt.Errorf("error scaling ekco operator back up: %w", err)
+	}
+	if err := scalePrometheus(ctx, cli, 2); err != nil {
```
Comment (emosbaugh) [inline pkg/cli/longhorn.go:178]: does it make sense here to use the desired scale annotation? what if we make a decision in the future to use only one replica for single node installations?

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: addons/rook/template/base/install.sh [hunk 3]
```diff
@@ -939,4 +939,15 @@ function rook_prompt_migrate_from_longhorn() {
     if ! confirmN; then
         bail "Not migrating"
     fi
+
+    local nodes=$(kubectl get nodes --no-headers | wc -l)
+    if [ "$nodes" -eq 1 ]; then
+        logFail "    ERROR: Your cluster has only one node, making Rook an unsuitable choice as a storage provisioner. It is strongly recommended that you install OpenEBS instead."
+        logFail "    Continuing with the Longhorn to Rook data migration under these conditions may result in unexpected errors potentially CAUSING DATA LOSS. Do you wish to proceed?"
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply addons/rook/template/base/install.sh]: Sorry, I think you reviewed an earlier version of the PR. I already have changed the message. Would you mind looking at it again ?

File: pkg/cli/longhorn.go [hunk 4]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
+			return nil, fmt.Errorf("error getting statefulset %s: %w", nsn, err)
+		}
+		return &sset, nil
+	case "ReplicaSet":
+		nsn := types.NamespacedName{Namespace: namespace, Name: owner.Name}
+		var rset appsv1.ReplicaSet
+		if err := cli.Get(ctx, nsn, &rset); err != nil {
+			return nil, fmt.Errorf("error getting replicaset %s: %w", nsn, err)
+		}
+		if len(rset.OwnerReferences) != 1 {
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:428]: Amazing, I was not aware of this helper. Adjusted.

File: pkg/cli/longhorn.go [hunk 5]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
 				log.Printf("$ kurl longhorn rollback-migration-replicas")
 				log.Printf("")
 			}
-			log.Print("All Longhorn volumes and nodes are healthy, ready for migration.")
+			log.Print("All Longhorn volumes and nodes are healthy.")
+
+			if err := scaleDownPodsUsingLonghorn(cmd.Context(), cli); err != nil {
+				return fmt.Errorf("error scaling down pods using longhorn volumes: %s", err)
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:164]: fixed.

File: pkg/cli/longhorn.go [hunk 6]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
+		return fmt.Errorf("error scaling ekco operator back up: %w", err)
+	}
+	if err := scalePrometheus(ctx, cli, 2); err != nil {
+		return fmt.Errorf("error scaling prometheus back up: %w", err)
+	}
+
+	log.Print("Scaling up pods using Longhorn volumes.")
+	var deps appsv1.DeploymentList
+	if err := cli.List(ctx, &deps); err != nil {
+		return fmt.Errorf("error listing longhorn deployments: %s", err)
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:185]: fixed.

File: pkg/cli/longhorn.go [hunk 7]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
+	}
+	for _, dep := range deps.Items {
+		if _, ok := dep.Annotations[pvmigrateScaleDownAnnotation]; !ok {
+			continue
+		}
+		replicas, err := strconv.Atoi(dep.Annotations[pvmigrateScaleDownAnnotation])
+		if err != nil {
+			return fmt.Errorf("error parsing replica count for deployment %s/%s: %s", dep.Namespace, dep.Name, err)
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:193]: fixed.

File: pkg/cli/longhorn.go [hunk 8]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
+		}
+		dep.Spec.Replicas = pointer.Int32(int32(replicas))
+		delete(dep.Annotations, pvmigrateScaleDownAnnotation)
+		log.Printf("Scaling up deployment %s/%s", dep.Namespace, dep.Name)
+		if err := cli.Update(ctx, &dep); err != nil {
+			return fmt.Errorf("error scaling up deployment %s/%s: %s", dep.Namespace, dep.Name, err)
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:199]: fixed.

File: pkg/cli/longhorn.go [hunk 9]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
+		}
+	}
+
+	var sts appsv1.StatefulSetList
+	if err := cli.List(ctx, &sts); err != nil {
+		return fmt.Errorf("error listing longhorn statefulsets: %s", err)
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:205]: fixed.

File: pkg/cli/longhorn.go [hunk 10]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
+	}
+	for _, st := range sts.Items {
+		if _, ok := st.Annotations[pvmigrateScaleDownAnnotation]; !ok {
+			continue
+		}
+		replicas, err := strconv.Atoi(st.Annotations[pvmigrateScaleDownAnnotation])
+		if err != nil {
+			return fmt.Errorf("error parsing replica count for statefulset %s/%s: %s", st.Namespace, st.Name, err)
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:213]: fixed.

File: pkg/cli/longhorn.go [hunk 11]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
+		}
+		st.Spec.Replicas = pointer.Int32(int32(replicas))
+		delete(st.Annotations, pvmigrateScaleDownAnnotation)
+		log.Printf("Scaling up statefulset %s/%s", st.Namespace, st.Name)
+		if err := cli.Update(ctx, &st); err != nil {
+			return fmt.Errorf("error scaling up statefulset %s/%s: %s", st.Namespace, st.Name, err)
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:219]: fixed.

File: pkg/cli/longhorn.go [hunk 12]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
+// and replicasets are supported (as those are the only types supported by pvmigrate).
+func getObjectsUsingLonghorn(ctx context.Context, cli client.Client) ([]client.Object, error) {
+	pods, err := getPodsUsingLonghorn(ctx, cli)
+	if err != nil {
+		return nil, fmt.Errorf("error getting pods using longhorn: %w", err)
+	}
+	seen := map[string]bool{}
+	var objects []client.Object
+	for _, pod := range pods {
+		if len(pod.OwnerReferences) == 0 {
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:389]: fixed.

File: pkg/cli/longhorn.go [hunk 13]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
+	); err != nil {
+		return fmt.Errorf("error waiting for ekco operator to scale down: %w", err)
+	}
+	return nil
+}
+
+// waitForPodsToBeScaledDown waits for all pods using matching the provided selector to disappear in the provided
+// namespace.
+func waitForPodsToBeScaledDown(ctx context.Context, cli client.Client, ns string, sel labels.Selector) error {
+	// return wait.PollImmediate(3*time.Second, 5*time.Minute, func() (bool, error) {
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:319]: this is what happens when you say "let me just do this dirty hack here so I can make it fail faster". missed the rollback, thanks.

File: pkg/cli/longhorn.go (hunk 1 above)
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:249]: fixed.

File: pkg/cli/longhorn.go [hunk 14]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
+			return fmt.Errorf("error scaling up statefulset %s/%s: %s", st.Namespace, st.Name, err)
+		}
+	}
+
+	log.Print("Pods using Longhorn volumes have been scaled up.")
+	return nil
+}
+
+// scaleDownPodsUsingLonghorn scales down all pods using Longhorn volumes.
+func scaleDownPodsUsingLonghorn(ctx context.Context, cli client.Client) error {
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:228]: Oh, we should definitely unify this. The problem I see here, and I have introduced this problem, is that I have used the `controller-runtime` client and the code in `flexvolume.go` uses the client-go's `Interface`.  Should I convert this code to also use client-go's Interface ? That would certainly involve some extra work and given the urgency on the task at hand I am not sure we should do it right now. Nonetheless this is worthy a discussion with the team about what client we use and then I can stop bringing chaos to the org.

File: pkg/cli/longhorn.go [hunk 15]
```diff
@@ -139,12 +158,377 @@ func NewLonghornPrepareForMigration(_ CLI) *cobra.Command {
...
+		nsn := types.NamespacedName{Namespace: prometheusNamespace, Name: prometheusName}
+		var prometheus promv1.Prometheus
+		if err := cli.Get(ctx, nsn, &prometheus); err != nil {
+			if errors.IsNotFound(err) {
+				return true, nil
+			}
+			return false, fmt.Errorf("error getting prometheus: %w", err)
+		}
+		prometheus.Spec.Replicas = &replicas
+		if err := cli.Update(ctx, &prometheus); err != nil {
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go:261]: That is a great idea. For sake of knowledge: I have tested this code using the oldest Prometheus version we have available and it was working. Fixed.

File: pkg/cli/longhorn.go (hunk 2 above)
Comment (ricardomaraschini) [reply pkg/cli/longhorn.go]: Fixed.

File: pkg/cli/longhorn.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (emosbaugh) [reply pkg/cli/longhorn.go:229]: i dont think it is required that we unify code given the urgency

File: scripts/common/longhorn.sh [hunk 16]
```diff
@@ -207,7 +215,14 @@ function longhorn_prepare_for_migration() {
     logFail "During the preparation for Longhorn, some replicas may have been scaled down to 0. Would you like to"
     logFail "restore the system to its original state?"
     if confirmY; then
-        "$DIR"/bin/kurl longhorn rollback-migration-replicas
+        log "Restoring Longhorn replicas to their original state"
+        longhorn_restore_replicas
```
Comment (emosbaugh) [inline scripts/common/longhorn.sh:219]:
```suggestion
        longhorn_restore_migration_replicas
```

File: scripts/common/longhorn.sh [hunk 17]
```diff
@@ -53,6 +53,33 @@ function longhorn_install_nfs_utils_if_missing_common() {
     fi
 }
 
+# longhorn_run_pvmigrate calls pvmigrate to migrate longhorn data to a different storage class. if a failure happen
+# then rolls back the original number of volumes and replicas.
+function longhorn_run_pvmigrate() {
+    local longhorn_sc=$1
+    local destStorageClass=$2
```
Comment (emosbaugh) [inline scripts/common/longhorn.sh:60]: im not a huge fan of mixing cases

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: scripts/common/longhorn.sh (hunk 16 above)
Comment (ricardomaraschini) [reply scripts/common/longhorn.sh]: Oh my, I can't look at this code anymore.

File: scripts/common/longhorn.sh (hunk 17 above)
Comment (ricardomaraschini) [reply scripts/common/longhorn.sh:60]: Yeah, me too. This was a copy and pasta from the Rook to OpenEBS code. Fixed.

**Review by emosbaugh — DISMISSED** [review]

**Comment by ricardomaraschini** [comment]
This one is going to be replaced by other two PRs I am raising.

**Pull request closed without merging**